### PR TITLE
Specify userland proxy path

### PR DESCRIPTION
**Pull request by runcom**
**\- What I did**

I've added a new daemon flag called `--userland-proxy-path` to configure the userland proxy path instead of having libnetwork to only rely on `docker-proxy` in `$PATH`. This is how other flags are currently working (`--add-runtime`, `--containerd` with a socket, `--runtime` in containerd etc etc) w/o defining an hardcoded string in a dependency (`docker-proxy` in libnetwork pkg).
This will also allow distributions to ship the binaries wherever they want instead of putting them indiscriminately in `$PATH` (we can already achieve this with all the other docker's binaries).

**\- How I did it**

vim

**\- How to verify it**
- `mv /usr/bin/docker-proxy /home/$USER/docker-proxy`
- `dockerd --userland-proxy-path /home/$USER/docker-proxy`

**\- Description for the changelog**

Add a new daemon flag `--userland-proxy-path` to configure the userland proxy instead of using the hardcoded `docker-proxy` from `$PATH`.

**\- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: man/dockerd.8.md
Comment (justincormack) [inline man/dockerd.8.md]: This is duplicated not a description

**Comment by justincormack** [comment]
I think this makes sense.

**Comment by justincormack** [comment]
Some docs changes in https://docs.docker.com/engine/installation/binaries/ to say they don't need to be in PATH might be useful.

**Comment by thaJeztah** [comment]
We should probably also support this in the `daemon.json` configuration file

File: vendor/src/github.com/docker/libnetwork/drivers/bridge/bridge.go
Comment (aboch) [inline vendor/src/github.com/docker/libnetwork/drivers/bridge/bridge.go]: Please do not add this new parameter in all the function calls.
Just modify portmapper constructor to take theproxypath.

File: man/dockerd.8.md
Comment (runcom) [inline man/dockerd.8.md]: Ack

File: vendor/src/github.com/docker/libnetwork/drivers/bridge/bridge.go
Comment (runcom) [inline vendor/src/github.com/docker/libnetwork/drivers/bridge/bridge.go]: Ack will fix

**Comment by runcom** [comment]
@aboch @justincormack took care of your comments, PTAL

**Comment by thaJeztah** [comment]
@runcom did you have a look at https://github.com/docker/docker/pull/26882#issuecomment-249370949 as well? Given that the userland-proxy can be enabled/disabled through `daemon.json`, I think this setting should be configurable through that as well

**Comment by runcom** [comment]
> @runcom did you have a look at #26882 (comment) as well? Given that the userland-proxy can be enabled/disabled through daemon.json, I think this setting should be configurable through that as well

@thaJeztah I did, I believe it can be already configured in `daemon.json` and no other changes are needed here (other than adding the new option in the config and defining the json name)

**Comment by thaJeztah** [comment]
Thanks for looking, @runcom, guess only documentation changes are needed then (add it to the daemon.json examples)

**Comment by runcom** [comment]
@thaJeztah added and also created https://github.com/docker/libnetwork/issues/1470

**Comment by runcom** [comment]
CI seems busted - e.g. no `vendor` build and janky's failing on vendors...

File: docs/reference/commandline/dockerd.md
Comment (justincormack) [inline docs/reference/commandline/dockerd.md]: the values here are the defaults, so this should be `""` I think.

File: man/dockerd.8.md
Comment (justincormack) [inline man/dockerd.8.md]: I think it is better to give an example like `/usr/local/bin/docker-proxy` as people may be unsure if it means path to find it or full path with binary name.

File: docs/reference/commandline/dockerd.md
Comment (thaJeztah) [inline docs/reference/commandline/dockerd.md]: I think it's ok to use this for illustration. The runtime path below also contains an example ("/usr/local/bin/my-runc-replacement")

File: man/dockerd.8.md
Comment (runcom) [inline man/dockerd.8.md]: > I think it is better to give an example like /usr/local/bin/docker-proxy as people may be unsure if it means path to find it or full path with binary name

the other _path_-like flags do not specify what you would like to see here - I guess it's implied it's a full path. Can you tell me how you'd like to see this sentence?

File: docs/reference/commandline/dockerd.md
Comment (justincormack) [inline docs/reference/commandline/dockerd.md]: ok, in which case I think `/usr/libexec/docker-proxy` is best as this would be the most common use case - not many people are going to override the proxy.

File: man/dockerd.8.md
Comment (justincormack) [inline man/dockerd.8.md]: "Specify the full filename of the userland proxy binary, instead of looking for `docker-proxy` in the `$PATH`"

Comment (runcom) [inline man/dockerd.8.md]: @justincormack fixed

File: docs/reference/commandline/dockerd.md
Comment (runcom) [inline docs/reference/commandline/dockerd.md]: fixed

**Comment by justincormack** [comment]
Looks good now, once sort out CI and vendoring.

**Comment by LK4D4** [comment]
Flaky test on janky. Unrelated.

**Comment by LK4D4** [comment]
@runcom btw I wonder if it worth to test this. Should be quite easy, at least I think so :)

**Comment by runcom** [comment]
@LK4D4, I'll write some tests later today, should be easy to mv the proxy binary and test it with -p

**Comment by LK4D4** [comment]
@runcom Yup, that's what I thought. Thanks!

**Comment by runcom** [comment]
@LK4D4 added an integration test

**Comment by runcom** [comment]
This is blocked by https://github.com/docker/docker/pull/26860 - it needs a libnetwork revendor there and one after into this PR to include https://github.com/docker/libnetwork/pull/1470

**Comment by runcom** [comment]
@LK4D4 @justincormack rebased, this should be good to go after CI

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Review by cpuguy83 — APPROVED** [review]
LGTM, moving to docs.

Need to add option to:
https://github.com/docker/docker/blob/master/docs/reference/commandline/dockerd.md#linux-configuration-file

**Comment by runcom** [comment]
@cpuguy83 I've already changed that https://github.com/docker/docker/pull/26882/files#diff-ca4f406de134be5cca3ae377c03a4386R1152 - haven't I?

**Comment by cpuguy83** [comment]
I guess I scrolled right past it. Only saw the man file update.

**Comment by runcom** [comment]
Ping @vdemeester @thaJeztah for doc review

**Comment by runcom** [comment]
:angel:

File: daemon/config_unix.go
Comment (thaJeztah) [inline daemon/config_unix.go]: Should we use the same description as is used in https://github.com/docker/docker/pull/26941 ?

```
Path to the userland proxy binary
```

File: man/dockerd.8.md
Comment (thaJeztah) [inline man/dockerd.8.md]: unrelated white-space change? or was this deliberate?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Comment by thaJeztah** [comment]
@runcom 🇩🇪 :smile:

File: daemon/config_unix.go
Comment (runcom) [inline daemon/config_unix.go]: makes sense,fixing

File: man/dockerd.8.md
Comment (runcom) [inline man/dockerd.8.md]: probably unrelated, I'll fix that I guess.

File: docs/reference/commandline/dockerd.md
Comment (thaJeztah) [inline docs/reference/commandline/dockerd.md]: Oh, just realized the flag itself is not shown in this file (usage output at the top)

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: docs/reference/commandline/dockerd.md
Comment (runcom) [inline docs/reference/commandline/dockerd.md]: I'll add that (if I've understood correctly your point lol)

**Comment by runcom** [comment]
@thaJeztah took care of your comments :tada:

**Review by thaJeztah — APPROVED** [review]
LGTM!

**Comment by thaJeztah** [comment]
ping @vdemeester for docs/merge

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Pull request merged**
